### DISCUSSION TITLE
Error reading netCDF file GridStat

### DISCUSSION THREAD

**Discussion by jtv108-wx** [dtcenter/METplus#2511] [File I/O] [answered]
Hello,

I'm having issues reading a netCDF file into GridStat. I'm getting this error in my output file,... 

--- NcCfFile::get_grid_from_lat_lon_vars() -> lon[367]=-112.672 lon[368]=-112.644  0.00133514 > 0.00133324
--- NcCfFile::get_grid_from_lat_lon_vars() -> MET can only process Latitude/Longitude files where the longitudes are evenly spaced (dlon=0.0266647, delta[368]=0.0279999)
--- NcCfFile::get_grid_from_lat_lon_vars() -> Please check the input data is the lat/lon projection

I believed the issue originates from the original netCDF file. From searching other discussion threads I believed the issue was similar to this (https://github.com/dtcenter/METplus/discussions/1931). I proceeded to writing a python script to manually read the original netCDF file and set the correct file attributes. Then I ran pcp_combine with python_embedding (using the python code) using the following command...

--- pcp_combine -add PYTHON_NUMPY $Output_filename -field 'name="$Path_to_ascii_numpy_script.py $Path_to_nc_file $nc_file_var";'

The pcp_combine command above successfully runs and outputs a netCDF file. However, when I pass this output netCDF file into GridStat I get the evenly spaced lat/lon error (see above).

Any assistance would be greatly appreciated!

Thanks.

**Comment by CPKalb** [comment]
Hi:

It sounds like the error is occurring when reading the output from Pcp-Combine into Grid-Stat?  Would you be able to send a sample netCDF file so that we could take a closer look at this file?  Instructions for sending data can be found [here](https://github.com/dtcenter/METplus/discussions/954).

Thanks,
Christina

**Reply by jtv108-wx** [reply]
Okay, great. I've sent a single sample netCDF file (unaltered from Pcp-Combine or Grid-Stat). Sorry for the large file, but I am trying to read the refc variable from the file.

Thanks,
Jake

**Reply by CPKalb** [reply]
Hi Jake:

I'm guessing this is the original file?  From your message above, it sounded like the error is occurring when trying to read the output from Pcp-Combine into Grid-Stat.  Is that correct?  If so, I would need to see the output file from Pcp-Combine to further debug.

Or, is the error occurring when you are trying to read the file sent into Grid-Stat by python embedding or directly?

Thanks,
Christina

**Reply by jtv108-wx** [reply]
Yes, that was the original file. Sorry for the confusion. I was hoping I may have missed something when dealing with the original file that would just let me read the original file directly into Grid-Stat (without the Pcp-Combine step needed). But yes, the "evenly spaced lat/lon" error I was getting was a result of reading the Pcp-combine (with python embedding) output into Grid-Stat. I'll send the Pcp-combine output file over right now.

Thanks,
Jake

**Reply by CPKalb** [reply]
Ah, I didn't realize you were wanting to omit the Pcp-Combine step.  Would you be able to post the python embedding script that you are using for Pcp-Combine here?  You will likely need to give it a .txt extension to be able to post.  Alternatively, you could put it on the ftp site.  

Thanks,
Christina

**Reply by jtv108-wx** [reply]
Well if it is not necessary to be able to read the file (into Grid-Stat) I wouldn't have to use the Pcp-Combine step. I was trying the fix I found here (https://github.com/dtcenter/METplus/discussions/1931) but I'm not sure if that actually applies to my case or files. But okay, I'll just add the python embedding script to the ftp site.

Thanks,
Jake

**Reply by CPKalb** [reply]
In the issue referenced, they used Pcp-Combine specifically because they didn't want to use python embedding with Grid-Stat (they were getting inconsistent VL1L2 output with python embedding due to a bug we had in MET).  If you are okay with reading the file into Grid-Stat using python embedding, then the Pcp-Combine step should not be necessary here.

The original file isn't in one of the [netCDF formats](https://met.readthedocs.io/en/latest/Users_Guide/data_io.html#input-data-formats) that MET can read directly.

Christina

**Reply by jtv108-wx** [reply]
Ah okay, sounds great. Yes, if possible I'd like to read the original netCDF file right into Grid-Stat (with python embedding). I'm very new to the python embedding component with MET/METplus.

Jake

**Comment by CPKalb** [comment, marked as answer]
Hi Jake:

I am able to get the data file to go directly into Grid-Stat using python embedding with MET version 11.1.0.  Are you running this using a METplus configuration file or directly in MET?  I tested directly in MET.  Assuming this is your forecast data, you would want to set the following to get it to work:

In the MET configuration file, for the forecast variable set: name       = "read_ascii_numpy.py 01_test.nc refc";
The level variable is omitted here in the configuration.  When you run Grid-Stat, you would want to do the following on the command line:
grid_stat PYTHON_NUMPY observation_input_file GridStatConfigurationFile

If you are running from METplus, the setup would be a bit different, and I would need to see your METplus configuration to help get it up and running.

When using python embedding, it's always a good idea to run a test using plot_data_plane to see if the data is correctly read into MET.  I did this using the following command: 
plot_data_plane PYTHON_NUMPY test_nc.ps 'name="read_ascii_numpy.py 01_test.nc refc";'

Here is the resulting image:
![test_nc](https://github.com/dtcenter/METplus/assets/22180338/7e2b0ab0-43c4-4843-a135-10791e2ebb0f)

The image created by MET's plot_data_plane is flipped compared to what I see when using ncview, which looks like the following:
<img width="579" alt="Screenshot" src="https://github.com/dtcenter/METplus/assets/22180338/7b679d79-9f8d-4f99-a776-c5163c4d5cc2">

As you can see, in ncview, the precipitation is in the upper right corner of the image whereas in the plot_data_plane output, it's in the lower right.  If the ncview version is correct, then you would want to change line 27 in your script to have the following 2 lines:
data = input_nc.variables[data_name][:]
met_data = data[::-1].copy()

With that change, I get an output image that matches what is displayed in ncview:
![test_nc_flipped](https://github.com/dtcenter/METplus/assets/22180338/87d4cf8b-de93-48eb-8c84-653aa2ff8e60)

Also, I think the date format you have in the python embedding script may be incorrect.  If the date of your data is May 1, 2023, then the init and valid should look like this:
'valid': '20230501_010000',
'init':  '20230501_000000',

I've attached the python embedding script here with the changes I made.
[read_ascii_numpy.py.txt](https://github.com/dtcenter/METplus/files/14530627/read_ascii_numpy.py.txt)

Christina

**Comment by jtv108-wx** [comment]
Thank you for the help. I am wondering if there is also an issue with the lat/lon data being read by MET as I'm fairly sure the data is suppose to be over the ECONUS (maybe a -360 type issue?). 

I am planning on running from METplus. I moved the conf file I use over to you to try. The command I use to run METplus is...

/usr/bin/python3.6 /usr/local/met10/METplus-4.0.0/ush/run_metplus.py -c $PATH_to_conf_file -c _$PATH_to_metplus_user.conf

Okay, I'll make those grid flipping changes once I confirm the location of the grid over the CONUS. Ah, you are correct on the date format issue, I'll make that change as well.

Jake

**Reply by CPKalb** [reply]
Hi Jake:

[This](https://metplus.readthedocs.io/en/latest/generated/met_tool_wrapper/GridStat/GridStat_python_embedding.html) is an example of how to set up a METplus configuration file when using python embedding in Grid-Stat.  I took your configuration file and made the changes that I think need to be made for this to work.  However, I can't actually test the configuration since I don't have your observations and mask file.  You may need to give the full path to your python script or to the input netCDF file in FCST_VAR1_NAME, depending on where you are running METplus and where the data/script is located.
[REFC_TEST01_GridStat.conf.txt](https://github.com/dtcenter/METplus/files/14531547/REFC_TEST01_GridStat.conf.txt)

For the netCDF file, if the grid is supposed to be over the CONUS, then most likely something is not correct in the grid specification.

Christina

**Reply by CPKalb** [reply]
I notice in the netCDF file that it states "2023-08-23T14:14 GRIB to CDM+CF via cfgrib-0.9.10.1/ecCodes-2.22.1".  GRIB/GRIB2 files can be easily read into MET so if you have the originals, it may be easier just to read them in as GRIB files.

Christina

**Comment by jtv108-wx** [comment]
Hi Christina,

Thank you for all of your help. I'll try setting up the conf file changes you attached for the METPlus python embedding run (with Grid-Stat). Ah, okay I'll double the grid specifications as well.

Okay, I'll see if I can track down the original GRIB files and try those as well.

Jake

**Reply by CPKalb** [reply]
Hi Jake:
If you have even one GRIB file, it would help with the grid specification string.  

When I look in the netCDF file, the smallest longitude is about 251, while the largest is almost 299.  For latitude, the values range between about 22 and 52.  This would mean the file covers part of the CONUS, running from about Montana, Wyoming, Colorado, and New Mexico eastward.  

MET doesn't read these latitude and longitude values directly when using python embedding, but rather uses the grid specified in the variable attrs.  The lat_pin and lon_pin values specified are completely outside the min and max latitude and longitude values found in the file.  I'm definitely not an expert at specifying lambert conformal grids, but I think that would be something to check in the grid specification.

Christina

**Reply by jtv108-wx** [reply]
Hi Christina,

I was able to get the files read into Grid-Stat with python embedding using your suggestions. You are correct, I double checked the grid with the file creator and got the lat_pin, lon_pin values appropriately set. So the data now is over the ECONUS. I can update the sample file I sent earlier if you'd like?

I am noticing an issue however when plotting the OBS (MRMS data file) and FCST output from the Grid-Stat _pairs.nc file? As you mentioned in your last post, the data in my FCST file do need to be flipped using your suggestions of  "met_data = data[::-1].copy()". However, once flipped (refc_test01_invert img) the orientation between the OBS and FCST grid do not align?

[OBS field]

![refc_test01_obs](https://github.com/dtcenter/METplus/assets/133161085/b4dde1ce-defd-4984-89af-02395013766d)

[FCST field "flipped"]
![refc_test01_fcst_invert](https://github.com/dtcenter/METplus/assets/133161085/d80d1cd3-bbce-44ce-928e-e3bdf7928be5)

I'm guessing this still goes back to grid issues with the FCST netCDF file? Or is a solution I could apply to the read_ascii.py script?

Jake

**Reply by CPKalb** [reply]
Hi Jake:

I'm assuming that by "not align" you are referring to the fact that the forecast grid seems to be offset from the total domain (as evidenced by the line running through New York, Pennsylvania southward, and also the one running across the MI, MN, etc.)?  Does this same thing happen when the data is plotted using plot_data_plane?  That would help narrow down whether the issue is something with the grid in the python script, or whether it's something that is happening in Grid-Stat.  

If you have a chance to either update the sample file or post the updated grid specification here, I could run plot_data_plane which would help give an idea where the issue may be originating.

Christina

**Reply by jtv108-wx** [reply]
Yes, align is probably not the right word but it seems the general shape of the whole domain is inverted when compared to the OBS grid. I have not tried plotting with plot_data_plane yet, I'll give that a try.

I have updated the sample file (01_test_update.nc) if you'd like to take a look. The lat_pin, lon_pin values in this file are now correct.

Thanks,
Jake

**Reply by CPKalb** [reply]
Hi Jake:

I ran a test in plot_data_plane using the grid specified in the new file, and the data came out as below:
![test_nc_flipped](https://github.com/dtcenter/METplus/assets/22180338/482e42f3-7ab8-47a2-980d-96125ff932f0)

The forecast data is aligned with the grid boundaries, but I notice that my domain is shorter on the top and longer on the bottom.  In your output from Grid-Stat, the domain is longer on top and shorter at the bottom.  Do you know what the domain should look like?  Should the longer side be on the top or on the bottom? 

I also ran a test in Grid-Stat, however, I don't have your observation file or mask file.  When I run Grid-Stat without the observation or mask, the output looks like the following:
<img width="580" alt="Screenshot 2024-03-12 at 1 06 48 PM" src="https://github.com/dtcenter/METplus/assets/22180338/baaa7fc2-3ebe-4b36-8355-6d9740b7b539">

It seems that your grid specification is different from the one I have, which is causing the domains to look different, and likely the cause of the misalignment in domain edge. 

Christina

**Reply by CPKalb** [reply]
As a secondary check, I loaded the data file in IDV to try and view reflectivity with the United States as a background.  The resulting image looks like this:
![IDV_refc](https://github.com/dtcenter/METplus/assets/22180338/8e3e4976-b51d-4f52-ba34-1f7bdebaba42)

In comparison to the MET output, the areas of precipitation are farther North and East in the IDV image.  IDV domains don't always display exactly as they do in MET, but the precipitation should be in a similar location.  This suggests that the grid specification (lat_pin, lon_pin, lon_orient, scale_lat, etc.) may still be incorrect.  MET doesn't read the grid specification in the netCDF file when using python embedding.  So, if you know what these numbers should be, this can be fixed inside the read_ascii.py script. 

I tried setting lat_pin = 25.7478095952292 and lon_pin = 255.027346249833 in read_ascii.py, and this is the resulting image:
![test_nc_g](https://github.com/dtcenter/METplus/assets/22180338/fa5e2c7f-d3d2-44c4-b89c-8d0ca231d65e)

It better matches what is seen in IDV for precipitation.  However, these numbers are guesses.  It would be good to check all the numbers in the grid specification to verify that everything is correct.

Christina

**Reply by jtv108-wx** [reply]
Hi Christina,

Thanks for taking the time to look into this further. Let me double check the grid settings and get a few more sample plots. I'll also need to double check the values with the original file creator. Thanks again for all of the help.

Jake

**Reply by CPKalb** [reply]
Hi Jake:

Here is a bit more information on some of the lambert conformal variables which will hopefully help get the right grid specification.  lat_pin and lon_pin are the reference latitude and longitude.  These could be the lower-left corner, the midpoint, or another location.  The The x_pin and y_pin values define the offset from the lower-left corner to the location of the reference lat/lon values.  In your grid, x_pin and y_pin are set to 0, which means that lat_pin and lon_pin should be the lower left corner points of the grid.  For lambert conformal grids, the lower left point of the grid are not always the minimum latitude and longitude.  lon_orient is the orientation longitude, which is the meridian of longitude that’s parallel to one of the vertical grid directions. 

Christina

**Reply by CPKalb** [reply]
Hi Jake:

I'm following up on this discussion.  Were you able to get this working?

Chrisinta

**Comment by jtv108-wx** [comment]
Hi Christina,

Sorry for the very delayed response. This email got buried in my inbox and then I have been away due to an ongoing medical issue. But yes, I was able to get the files read and processed from METplus. Thank you so much for all of your help.

Thanks,
Jake
________________________________
